azurerm/internal/services/compute/resource_arm_disk_encryption_set.go

```diff
@@ -0,0 +1,387 @@
+		if err := d.Set("active_key", flattenArmDiskEncryptionSetKeyVaultAndKeyReference(encryptionSetProperties.ActiveKey)); err != nil {
+			return fmt.Errorf("Error setting `active_key`: %+v", err)
+		}
+		if err := d.Set("previous_keys", flattenArmDiskEncryptionSetKeyVaultAndKeyReferenceArray(encryptionSetProperties.PreviousKeys)); err != nil {
+			return fmt.Errorf("Error setting `previous_keys`: %+v", err)
+		}
+	}
+	if identity := resp.Identity; identity != nil {
+		if err := d.Set("identity", flattenArmDiskEncryptionSetIdentity(identity)); err != nil {
+			return fmt.Errorf("Error setting `identity`: %+v", err)
+		}
+	}
+
+	return tags.FlattenAndSet(d, resp.Tags)
+}
+
+func resourceArmDiskEncryptionSetDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Compute.DiskEncryptionSetsClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["diskEncryptionSets"]
+
+	future, err := client.Delete(ctx, resourceGroup, name)
+	if err != nil {
+		if response.WasNotFound(future.Response()) {
+			return nil
+		}
+		return fmt.Errorf("Error deleting Disk Encryption Set %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("Error waiting for deleting Disk Encryption Set %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	}
+
+	return nil
+}
+
+func expandArmDiskEncryptionSetKeyVaultAndKeyReference(input []interface{}) *compute.KeyVaultAndKeyReference {
+	if len(input) == 0 {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+
+	sourceVaultId := v["source_vault_id"].(string)
+	keyUrl := v["key_url"].(string)
+
+	result := compute.KeyVaultAndKeyReference{
+		KeyURL: utils.String(keyUrl),
+		SourceVault: &compute.SourceVault{
+			ID: utils.String(sourceVaultId),
+		},
+	}
+	return &result
+}
+
+func expandArmDiskEncryptionSetIdentity(input []interface{}) *compute.EncryptionSetIdentity {
+	if len(input) == 0 {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+
+	t := v["type"].(string)
+	result := compute.EncryptionSetIdentity{
+		Type: compute.DiskEncryptionSetIdentityType(t),
+	}
+
+	return &result
+}
+
+func flattenArmDiskEncryptionSetKeyVaultAndKeyReference(input *compute.KeyVaultAndKeyReference) []interface{} {
+	if input == nil {
+		return make([]interface{}, 0)
+	}
+
+	result := make(map[string]interface{})
+
+	if keyUrl := input.KeyURL; keyUrl != nil {
+		result["key_url"] = *keyUrl
+	}
+	if sourceVault := input.SourceVault; sourceVault != nil {
+		if sourceVaultId := sourceVault.ID; sourceVaultId != nil {
+			result["source_vault_id"] = *sourceVaultId
+		}
+	}
+
+	return []interface{}{result}
```

Comment:
both of these fields need to always have a value set:

```suggestion
	keyUrl := ""
	if input.KeyURL != nil {
		keyUrl = *input.KeyURL
	}
	
	sourceVaultId := ""
	if input.SourceVault != nil && input.SourceVault.ID != nil {
		sourceVaultId = *input.SourceVault.ID
	}

	return []interface{}{
		map[string]interface{}{
			"key_url": keyUrl,
			"source_vault_id": sourceVaultId,
		},
	}
```